src/librustc_driver/args/mod.rs

```diff
@@ -0,0 +1,207 @@
+/// @file - Read command-line options from file. The options read are inserted in place of the
+/// original @file option. If file does not exist, or cannot be read, then the option will be
+/// treated literally, and not removed.
+///
+/// Options in file are separated by whitespace. A whitespace character may be included in an
+/// option by surrounding the entire option in either single or double quotes. Any character
+/// (including a backslash) may be included by prefixing the character to be included with a
+/// backslash. The file may itself contain additional @file options; any such options will be
+/// processed recursively.
+///
+/// See https://sourceware.org/binutils/docs/ld/Options.html#Options for the documentation of
+/// gnu-ld's @file parameters.
+pub struct ArgsStack {
+    base: env::ArgsOs,
+    stack: Vec<Box<dyn Iterator<Item = FileArg>>>,
+}
+
+impl ArgsStack {
+    pub fn new() -> Self {
+        ArgsStack { base: env::args_os(), stack: vec![] }
+    }
+
+    fn stack_next(&mut self) -> Option<FileArg> {
+        while let Some(top) = self.stack.last_mut() {
+            match top.next() {
+                Some(filearg) => return Some(filearg),
+                None => {
+                    let _ = self.stack.pop();
+                }
+            }
+        }
+
+        None
+    }
+}
+
+impl Iterator for ArgsStack {
+    type Item = OsString;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            let arg = self.stack_next().or_else(|| self.base.next().map(FileArg::from));
+            match arg {
+                None => return None,
+                Some(FileArg::Arg(arg)) => return Some(arg),
+                Some(FileArg::Path(path)) => {
+                    let file = if self.stack.len() >= MAX_RECURSION {
+                        eprintln!(
+                            "Warning: Max recursion depth {} reached for file {}",
+                            MAX_RECURSION,
+                            path.display()
+                        );
+                        None
+                    } else {
+                        match fs::read(&path) {
+                            Ok(file) => Some(file),
+                            Err(err) => {
+                                eprintln!(
+                                    "Warning: Failed to open {}: {} (cwd: {})",
+                                    path.display(),
+                                    err,
+                                    env::current_dir().unwrap().display(),
+                                );
+                                None
+                            }
+                        }
+                    };
+                    if let Some(file) = file {
+                        self.stack.push(Box::new(file_args(file)))
+                    } else {
+                        // Failed to open - try to reconstruct original arg
+                        let arg = format!("@{}", path.display());
```

Comment:
This sort of seems like odd behavior to me to inherit in rustc. It seems like the only reason to have this would be if you're compiling `@foo.rs` as a file, but that seems so rare it feels reasonable to me to just say we can't compile files that look like that and `@` is always reserved for arg-files.

Reply:
Yeah, and you can always do `./@foo.rs`.

So the alternatives are:
1. just omit it entirely, or
2. exit with an error

2 seems better.

Reply:
Ah yeah that's a good point, in that case I say we just unconditionally exit with an error because you can `./` your way to victory if necessary